### PR TITLE
[ASTS] Foreground Task Part IV: Decoupling Reader Timestamps

### PR DESCRIPTION
**Pull request by jeremyk-91**
## General
**Before this PR**:
There's an issue with #7270 where a cell written by a transaction that started in bucket A and committed in bucket B would not be considered for sweeping. (See https://github.com/palantir/atlasdb/pull/7288/files#r1785120409.) This is because
- in the current implementation of `DefaultSingleBucketSweepTask`, we set the sweep timestamp to the minimum of the actual sweep timestamp and the end of the bucket when calling the `SweepQueueReader`, and
- in `SweepQueueReader`, we use the sweep timestamp both for the purposes of 
  - 1️⃣ controlling our reads to `SweepableTimestamps` and `SweepableCells`, **and** 
  - 2️⃣ filtering out writes corresponding to transactions that started before the sweep timestamp but committed after, because values behind these writes might still be readable to live transactions.

Normally this is okay, because the sweep timestamp increases over time - while using the sweep timestamp for 1️⃣ will mean that we end up reading some values that haven't committed yet, we'll come back and retry later. However, if we clamp the sweep timestamp to the end of the bucket range, then for a cell written across a sweep boundary we will never succeed the 2️⃣ check (though it's important that we don't actually sweep it until the 2️⃣ check would pass).

**After this PR**:
We thus explicitly treat the bounding concepts for 1️⃣ and 2️⃣ separately in `SweepQueueReader`: the _exclusive max processable cell start timestamp_ (which is used for the purposes of 1️⃣), and the _sweep timestamp_ (which is used for the purposes of 2️⃣). Standard targeted sweep uses the sweep timestamp for both of these; ASTS uses the minimum of the bucket end range and the sweep timestamp for 1️⃣, and just the sweep timestamp for 2️⃣.

It's worth considering the two newly added integration tests, as these capture the semantics I was looking for here.

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:
- Please check that the splitting of the original usage of sweep timestamp into the individual 1️⃣ and 2️⃣ scenarios has been performed correctly.
- There were multiple points at which decisions could be made about calculating the minimum of the range end and sweep timestamp, as well as whether to throw if the timestamp for 1️⃣ was greater than that of 2️⃣. I went with the idea that the core Sweep classes (`SweepableCells`, `SweepQueueReader`) can make assumptions about how they're being used / can expect to only be used by classes directly manipulating Sweep persistence, and so they should validate the 1️⃣ <= 2️⃣ invariant, while the ASTS task is itself responsible for bounding things correctly. Does this make sense?
  - There's also a version of this where the `Reader` which is how one interfaces with the queues is responsible for doing the min with the sweep timestamp, though I didn't have a strong preference for either way.

**Is documentation needed?**: No.

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: No

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**: No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**: Yes

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**: No

**Does this PR need a schema migration?** No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**: Nothing in particular I think.

**What was existing testing like? What have you done to improve it?**: Added one new unit test plus two new integration tests that fail on the previous bad behaviour (plus generally added some unit tests at places in the code that were changed).

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**: I don't think this applies.

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**: N/A

## Execution
Does not change production.

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**: No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**: No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**: Not aware of anything obvious

## Development Process
**Where should we start reviewing?**: SingleBucketSweepTaskIntegrationTest

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**: It's not.

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableTimestamps.java [hunk 1]
```diff
@@ -73,13 +73,13 @@ private SweepableTimestampsTable.SweepableTimestampsColumn computeColumn(long st
...
-     * @param sweepTs exclusive maximum timestamp to check for
+     * @param maxTsExclusive exclusive maximum timestamp to check for
      * @return Optional containing the fine partition, or Optional.empty() if there are no more candidates before
      * sweepTs
      */
-    Optional<Long> nextTimestampPartition(ShardAndStrategy shardStrategy, long lastSweptTs, long sweepTs) {
+    Optional<Long> nextTimestampPartition(ShardAndStrategy shardStrategy, long lastSweptTs, long maxTsExclusive) {
         long minFineInclusive = SweepQueueUtils.tsPartitionFine(lastSweptTs + 1);
-        long maxFineInclusive = SweepQueueUtils.tsPartitionFine(sweepTs - 1);
+        long maxFineInclusive = SweepQueueUtils.tsPartitionFine(maxTsExclusive - 1);
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableTimestamps.java:82]: this looks random but was consistent with `SweepableCells` usage, as well as the javadoc comments.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java [hunk 2]
```diff
@@ -564,6 +564,73 @@ public void makesPartialProgressOnAnOpenBucket(SweepStrategyTestContext sweepStr
...
+    //                                 ^ Sweep Timestamp
+    //                     ^ Sweep Task
+    @ParameterizedTest
+    @MethodSource("testContexts")
+    public void successfullySweepsCellInMyBucketCommittingInAFutureBucket(
+            SweepStrategyTestContext sweepStrategyTestContext) {
+        when(completelyClosedSweepBucketBoundRetriever.getStrictUpperBoundForCompletelyClosedBuckets())
+                .thenReturn(2L);
+        writeCell(sweepStrategyTestContext.dataTable(), END_OF_BUCKET_ONE - 3, DEFAULT_VALUE, END_OF_BUCKET_ONE - 2);
+        writeCell(sweepStrategyTestContext.dataTable(), END_OF_BUCKET_ONE - 1, ANOTHER_VALUE, END_OF_BUCKET_ONE);
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java]: It would be marginally clearer if:
a) commitTs was END_OF_BUCKET_ONE + 1, or
b) we note that the end of bucketOne is an exclusive timestamp.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java [hunk 3]
```diff
@@ -564,6 +564,73 @@ public void makesPartialProgressOnAnOpenBucket(SweepStrategyTestContext sweepStr
...
+    @ParameterizedTest
+    @MethodSource("testContexts")
+    public void doesNotSweepCellInMyBucketCommittingInAFutureBucketIfSweepTimestampNotPastItsCommit(
+            SweepStrategyTestContext sweepStrategyTestContext) {
+        when(completelyClosedSweepBucketBoundRetriever.getStrictUpperBoundForCompletelyClosedBuckets())
+                .thenReturn(2L);
+        writeCell(sweepStrategyTestContext.dataTable(), END_OF_BUCKET_ONE - 30, DEFAULT_VALUE, END_OF_BUCKET_ONE - 20);
+        writeCell(sweepStrategyTestContext.dataTable(), END_OF_BUCKET_ONE - 10, ANOTHER_VALUE, END_OF_BUCKET_ONE + 1);
+
+        sweepTimestamp.set(END_OF_BUCKET_ONE);
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java]: consider: Should we still check that the first value has been swept away?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java [hunk 4]
```diff
@@ -145,12 +146,11 @@ private void markBucketCompleteIfEligible(SweepableBucket sweepableBucket) {
         }
     }
 
-    private static long getEndOfSweepRange(SweepableBucket sweepableBucket, long sweepTimestampForIteration) {
+    private static long getLastProcessableCellStartTimestamp(SweepableBucket sweepableBucket, long sweepTimestamp) {
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java]: nit: Is there a reason you chose last over max here? In SBA we refer to it as the maxProcessableCellStartTimestamp

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepBatchAccumulator.java [hunk 5]
```diff
@@ -39,22 +39,22 @@ class SweepBatchAccumulator {
     private boolean anyBatchesPresent = false;
     private boolean nextBatchAvailable = true;
 
-    SweepBatchAccumulator(long sweepTimestamp, int batchSizeThreshold, long progressTimestamp) {
-        this.sweepTimestamp = sweepTimestamp;
+    SweepBatchAccumulator(long maxProcessableCellStartTimestamp, int batchSizeThreshold, long progressTimestamp) {
+        this.maxProcessableCellStartTimestamp = maxProcessableCellStartTimestamp;
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepBatchAccumulator.java]: Can we make it clear this is exclusive? When I read this, I think inclusive (probably because of the word max?)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java [hunk 6]
```diff
@@ -163,8 +163,28 @@ private static long tsMod(long timestamp) {
 
     SweepBatch getBatchForPartition(
             ShardAndStrategy shardStrategy, long partitionFine, long minTsExclusive, long sweepTs) {
+        return getBatchForPartition(shardStrategy, partitionFine, minTsExclusive, sweepTs, sweepTs);
+    }
+
+    /**
+     * Variant of {@link #getBatchForPartition(ShardAndStrategy, long, long, long)}, that considers both a maximum
+     * start timestamp that we want to read from the table as well as the sweep timestamp separately. This is
+     * important for the sake of cells which started before maxStartTs but committed after that.
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java]: nit - personal preference, but if you're mentioning parameter names please reference them exactly! `maxStartTsExclusive`. To me, it makes it very easy to match them out without having to prefix match

**Review by mdaudali — APPROVED** [review]
I have checked that the changes are both necessary and sufficient. 
Some comments around naming, but non-blocking.

I have a question about `getBatchOfWrites` in SweepableCells, but we can chat about that offline - it's not relevant to the PR.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java]: Yep, I went with (b) throughout, as I think it's better to test the "true" boundary case.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java]: Hmm, actually the first value won't be swept away (a range tombstone is applied but it doesn't really cover anything, and there isn't a nice way to check this as far as I know, especially for thorough). I could do `if strategy is conservative then check a sentinel exists` but that seems a bit unpleasant - what do you think?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DefaultSingleBucketSweepTask.java]: I'll use max. Yeah, should be consistent. I think I had some conflicts about the naming because of inclusive/exclusive things (as you note later!)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepBatchAccumulator.java (hunk 5 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepBatchAccumulator.java]: Yep, will add.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java (hunk 6 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCells.java]: Makes sense, changed to the full form.
